tlair/src/main/scala/at/forsyte/apalache/tla/lir/TlaType1.scala

```diff
@@ -310,3 +313,112 @@ case class OperT1(args: Seq[TlaType1], res: TlaType1) extends TlaType1 {
   }
 
 }
+
+/**
+ * ****************** EXTENSIONS OF ADR014 ******************************
+ */
+
+/**
+ * A row type that contains a collection of fields with assigned types and an optional variable that represents the part
+ * to be defined.
+ *
+ * @param fieldTypes
+ *   a mapping from field names to field types, which may be parameterized
+ * @param other
+ *   the undefined part: Either None, or Some(variable)
+ */
+case class RowT1(fieldTypes: SortedMap[String, TlaType1], other: Option[VarT1]) extends TlaType1 {
+  override def toString: String = {
+    val pairs = fieldTypes.map(p => "%s: %s".format(p._1, p._2)).mkString(" | ")
+    other match {
+      case None    => if (pairs.nonEmpty) s"(| $pairs |)" else "(||)"
+      case Some(v) => s"(| $pairs | $v |)"
+    }
+  }
+
+  override def usedNames: Set[Int] =
+    fieldTypes.values
+      .map(_.usedNames)
+      .foldLeft(other.map(_.usedNames).getOrElse(Set.empty))(_ ++ _)
+}
+
+object RowT1 {
+  def apply(fieldTypes: (String, TlaType1)*): RowT1 = {
+    RowT1(SortedMap(fieldTypes: _*), None)
+  }
+
+  def apply(other: VarT1, fieldTypes: (String, TlaType1)*): RowT1 = {
+    RowT1(SortedMap(fieldTypes: _*), Some(other))
+  }
+}
+
+/**
+ * A record constructed from a row type. That the row is of the right kind should be checked by the kind unifier.
+ *
+ * @param row
+ *   RowNil, RowCons, or VarT1.
+ */
+case class RecRowT1(row: TlaType1) extends TlaType1 {
+  override def toString: String = {
+    row match {
+      case RowT1(fieldTypes, other) =>
+        // the special user-friendly form, e.g., { foo: Int, g: Bool } or { foo: Int, g: Bool, a }
+        val fields = fieldTypes.map(p => "%s: %s".format(p._1, p._2)).mkString(", ")
+        other match {
+          case None    => s"{ $fields }"
+          case Some(v) => s"{ $fields, $v }"
+        }
+
+      case _ =>
+        // the general case, which is probably never reachable
```

Comment:
I feel like making a distinction of kinds will help us here. `RecRowT1` and `VariantT1` should be constructed with parameters of kind `row`, and then we can make sure that `RowT1` (of kind row) is matched at scala's type system level.

As I see it, `TlaType1` is the only kind we have right now. We could introduce a `TlaRow = RowT1 | VariantT1` and remove this two types from `TlaType1`.

I see there is an "Implement kind inference?" at the meta-issue, and I think we should do it to avoid mixing things up. We can do it in another iteration, of course.

Reply:
Right. I do not have a clear opinion about kinds yet. `VariantT1` and `RecRowT1` are not rows, but they require rows as the arguments. The main reason for having `row: TlaType1` and not `row: RowT1` was the need for accommodating `VarT1`. However, we also have `VarT1` as an optional parameter of `RowT1` now. So maybe we can just require `row` to have the type `RowT1`?

Reply:
Yes, you're right! I guess that's good enough for now, I'm just concerned that we might have problems if `RowT1` is a `TlaType1` as it is right now. We shouldn't unify term types with rows - but we can figure out how to distinguish them when implementing unification for this.

I'll leave this image here for future reference because I keep going back to the paper to check it:
![image](https://user-images.githubusercontent.com/18356186/162449859-0a5885ca-1ef9-4674-af70-81648f9808c1.png)